tests/test_rotations.py

```diff
@@ -0,0 +1,110 @@
+    output[2] = (w1 * y2) - (x1 * z2) + (y1 * w2) + (z1 * x2)
+    output[3] = (w1 * z2) + (x1 * y2) - (y1 * x2) + (z1 * w2)
+    return output
+
+
+def rotate_vector_by_quaternions(coords: NDArray, quaternion: NDArray) -> NDArray:
+    """
+    Reference implementation of rotating coordinates using a quaternion
+
+    References
+    ----------
+    https://math.stackexchange.com/a/535223
+    """
+    assert coords.shape == (3,)
+    assert quaternion.shape[0] == 4
+    expanded_coords = np.concatenate([np.zeros(1), coords])
+    quaternion_conjugate = quaternion.copy()
+    # Compute quaternion conjugate by taking the conjugate of all but w
+    quaternion_conjugate[1:] *= -1
+    ham_coords = hamiliton_product(quaternion, expanded_coords)
+    return hamiliton_product(ham_coords, quaternion_conjugate)[1:]  # Truncate off w
+
+
+def convert_quaternion_for_scipy(quat: NDArray) -> NDArray:
+    """Scipy has the convention of (x, y, z, w) which is different than the wikipedia definition, swap ordering to verify using scipy.
+
+    References
+    ----------
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_quat.html
+    """
+    return np.append(quat[1:], [quat[0]])
+
+
+@pytest.mark.nogpu
+@pytest.mark.parametrize("seed", list(range(10)))
+@pytest.mark.parametrize("size", [1000])
+def test_reference_rotation_by_quaternion(seed, size):
+    rng = np.random.default_rng(seed)
+
+    all_coords = rng.normal(size=(size, 3)) * 100.0
+
+    quaternions = rng.normal(size=(size, 4))
+    # Normalize the quaternions to a vector quaternion, implicitly performed by Rotation
+    norms = np.linalg.norm(quaternions, axis=-1)
+    quaternions = quaternions / np.expand_dims(norms, -1)
+    for coords, quaternion in zip(all_coords, quaternions):
+
+        rotation = Rotation.from_quat(convert_quaternion_for_scipy(quaternion))
+
+        ref_coords = rotation.apply(coords)
+
+        test_coords = rotate_vector_by_quaternions(coords, quaternion)
+
+        np.testing.assert_allclose(ref_coords, test_coords)
+
+
+@pytest.mark.memcheck
+@pytest.mark.parametrize("seed", [2023])
+@pytest.mark.parametrize("precision,atol,rtol", [(np.float64, 1e-8, 1e-8), (np.float32, 2e-5, 1e-5)])
+@pytest.mark.parametrize("n_rotations", [2, 33, 100])
+@pytest.mark.parametrize("n_coords", [8, 33, 65])
+def test_cuda_rotation_by_quaternion(seed, precision, atol, rtol, n_rotations, n_coords):
+    rng = np.random.default_rng(seed)
+
+    all_coords = rng.normal(size=(n_coords, 3)) * 100.0
+
+    quaternions = rng.normal(size=(n_rotations, 4))
+    rotate_function = custom_ops.rotate_coords_f32
+    if precision == np.float64:
+        rotate_function = custom_ops.rotate_coords_f64
+
+    rotated_coords = rotate_function(all_coords, quaternions)
+
+    norms = np.linalg.norm(quaternions, axis=-1)
+    # Normalize the quaternions to a vector quaternion after passing to C++, as should be automatically normalized
```

Comment:
(think this should say "unit" here too?)
```suggestion
    # Normalize the quaternions to a unit quaternion after passing to C++, as should be automatically normalized
```